src/pages/en/core-concepts/endpoints.md

```diff
@@ -0,0 +1,275 @@
+      url: 'https://astro.build/',
+    }),
+  };
+}
+```
+
+The return object can also have an `encoding` property. It can be any valid [`BufferEncoding`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/bdd02508ddb5eebcf701fdb8ffd6e84eabf47885/types/node/buffer.d.ts#L169) accepted by node.js' `fs.writeFile` method. For example, to produce a binary png image:
+
+```ts title="src/pages/astro-logo.png.ts" {7}
+export async function get({ params, request }) => {
+  const response = await fetch("https://astro.build/assets/press/full-logo-light.png");
+  const buffer = Buffer.from(await response.arrayBuffer());
+  return {
+    body: buffer,
+    encoding: 'binary',
+  };
+```
+
+You can type your endpoint functions using the `APIRoute` type:
+
+```js
+import type { APIRoute } from 'astro';
+
+export const get: APIRoute = async function get ({params, request}) {
+...
+```
+
+As their only argument, endpoint functions recieve an object with two properties: `params` and `request`.
+
+### `params` and Dynamic routing
+
+Endpoints support the same [dynamic routing](/en/core-concepts/routing/#dynamic-routes) features that pages do. Name your file with a bracketed parameter name and export a [`getStaticPaths()` function](/en/reference/api-reference/#getstaticpaths). Then, you can access the parameter using the `params` property passed to the endpoint function:
+
+```ts title="src/pages/[id].json.ts"
+import type { APIRoute } from 'astro';
+
+const usernames = ["Sarah", "Chris", "Dan"]
+
+export const get: APIRoute = ({ params, request }) => {
+  const id = params.id;
+  return {
+    body: JSON.stringify({
+      name: usernames[id]
+    })
+  }
+};
+
+export function getStaticPaths () {
+    return [ 
+        { params: { id: "0"} },
+        { params: { id: "1"} },
+        { params: { id: "2"} },
+    ]
+}
+```
+
+This will generate three JSON endpoints at build time: `/api/1.json`, `/api/2.json`, `/api/3.json`. Dynamic routing with endpoints works the same as it does with pages, but because the endpoint is a function and not a component, [props](/en/reference/api-reference/#data-passing-with-props) aren't supported.
+
+### `request`
+All endpoints receive a `request` property, but only `request.url` works in static mode. This returns the full URL of the current endpoint, using your [`site`](/en/reference/configuration-reference/#site) config option as the base.
+
+```ts title="src/pages/request-path.json.ts"
+import type { APIRoute } from 'astro';
+
+export const get: APIRoute = ({ params, request }) => {
+  return {
+    body: JSON.stringify({
+      path: new URL(request.url).pathname
+    })
+  };
+}
+```
+
+## Server Endpoints (API Routes)
+Everything in the previous section can be used in SSR mode, but the endpoints will be built when they are requested. This unlocks new features that are unavailable at build time, and allows you to build API routes that listen for requests and securely execute code on the server at runtime.
```

Comment:
```suggestion
Everything described in the previous static files section can also be used in SSR mode: files export a `get` function and return a `body`, dynamic routing is available, and all endpoints receive a request property. But, unlike in `static` mode, when you configure `server` mode, the endpoints will be built when they are requested. This unlocks new features that are unavailable at build time, and allows you to build API routes that listen for requests and securely execute code on the server at runtime.
```
Love the second sentence here especially!

The concern I have is that some people are going to jump to this section immediately, not having read the from the top of the page ("I'm doing SSR!"). So, I think this section needs a bit more of an explicit lead in and comparison with what came before. Not totally married to the suggestion above, but something like that I think would make that hit home?